File: Java/LRU-Cache.java
/**
* LRU Cache
*
* Logic : maintain doubly linked List for insertion at head and deletion from tail and maintain       		  hashmap of key as key and value as node. Node consists of key, value, prev and next  	    		  pointers.
* Runtime: 16 ms
* Memory Usage: 47.9 MB
*/

class LRUCache {

    final Node head = new Node();
    final Node tail = new Node();
    int capacity;
    Map<Integer, Node> map;
    
    public LRUCache(int capacity) {
        map = new HashMap(capacity);
        this.capacity = capacity;
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        int result =  -1;
        Node node = map.get(key);
        if(node!=null)
        {
            remove(node);
            add(node);
            result = node.val;
        }
        return result;
    }
    
    public void put(int key, int value) {
        Node node = map.get(key);
        if(node!=null)
        {
            remove(node);
            node.val = value;
            add(node);
        }
        else{
            if(map.size() == capacity)
            {
                map.remove(tail.prev.key);
                remove(tail.prev);
            }
            Node new_node = new Node();
            
            new_node.key = key;
            new_node.val = value;
            map.put(key, new_node);
            add(new_node);
        }
    }
    
    public void add(Node node)
    {
        Node head_next = head.next;
        node.next = head_next;
        head_next.prev = node;
        head.next = node;
        node.prev = head;
        
    }
    
    public void remove(Node node)
    {
        Node next_node = node.next;
        Node prev_node = node.prev;
        
        next_node.prev = prev_node;
        prev_node.next = next_node;
        
    }
    
    class Node{
        int key;
        int val;
        Node prev;
        Node next;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: Python/LRUCache.py
class ListNode(object):
    def __init__(self,key,val):
        self.key = key
        self.value  = val
        self.next = None
        self.prev  = None

class LRUCache:

    def __init__(self, capacity: int):
        self.cap = capacity
        self.dic = {}
        self.head = ListNode(-1,-1)
        self.tail = ListNode(-1,-1)
        self.head.next = self.tail
        self.tail.prev = self.head
        

    def get(self, key: int) -> int:
        if key in self.dic:
            node = self.dic[key]
            self.remove(node)
            self.add(node)
            return node.value
        return -1

    def put(self, key: int, value: int) -> None:
    
        if key in self.dic:
            node = self.dic[key]
            self.remove(node)
            del self.dic[node.key]
        if len(self.dic)>=self.cap:
            node = self.head.next
            self.remove(node)
            del self.dic[node.key]
            
        node = ListNode(key,value)
        self.dic[key] = node
        self.add(node)
        # self.show(self.head)
        
        
    
    def remove(self,node):
        prevNode = node.prev
        prevNode.next = node.next
        node.next.prev = prevNode
        
    def add(self,node):
        tailprev = self.tail.prev
        tailprev.next = node
        node.prev = tailprev
        node.next = self.tail
        self.tail.prev = node
        # self.show(self.head)
        
    def show(self,head):
        while head:
            print(head.value,end=" ")
            head = head.next
        print()
            


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
